### PR TITLE
More json refactoring towards System.Text.Json

### PR DESCRIPTION
**Pull request by simeoncran**
This change has been tested and does not change the output of the corpus.
The code is a bit more resilient (i.e. less likely to throw exceptions if there's unexpected input).
I expect there will be 2 more PRs for us to actually get switched over to System.Text.Json.

### CODE REVIEW

File: source/LottieReader/LottieReader.projitems [hunk 1]
```diff
@@ -13,17 +13,15 @@
     <Compile Include="$(MSBuildThisFileDirectory)Serialization\Enums.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Serialization\Exceptions.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)Serialization\Fonts.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)Serialization\JCArray.cs" />
-    <Compile Include="$(MSBuildThisFileDirectory)Serialization\JCObject.cs" />
```
Comment (derobinsMSFT) [inline source/LottieReader/LottieReader.projitems]: >JCObject.cs" [](start = 63, length = 12)

Why did this get removed here? The file wasn't deleted.

Comment (simeoncran) [reply source/LottieReader/LottieReader.projitems]: Good find. Gone now.

---
In reply to: [380877959](https://github.com/windows-toolkit/Lottie-Windows/pull/244#discussion_r380877959) [](ancestors = 380877959)

File: LottieGen/LottieGen.csproj [hunk 2]
```diff
@@ -23,14 +23,14 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
     <PackageReference Include="StyleCop.Analyzers" Version="1.1.1-beta.61">
```
Comment (simeoncran) [inline LottieGen/LottieGen.csproj]: Should be able to remove this - there's one of these in Directory.Build.Props already.

File: source/LottieReader/Serialization/ShapeLayerContents.cs [hunk 3]
```diff
@@ -54,68 +54,70 @@ ShapeLayerContent ReadShapeContent(JCObject obj)
...
-
-            var numberOfProperties = ReadInt(obj, "np");
-            var items = ReadShapesList(obj.GetNamedArray("it", null));
-            AssertAllFieldsRead(obj);
+            // Not clear whether we need to read these properties.
+            obj.IgnorePropertyThatIsNotYetSupported("cix", "cl", "ix", "hd");
+
+            var numberOfProperties = obj.Int32OrNullProperty("np");
+            var items = ReadShapesList(obj.AsArrayProperty("it"));
+            obj.AssertAllPropertiesRead();
```
Comment (simeoncran) [inline source/LottieReader/Serialization/ShapeLayerContents.cs]: >  obj.AssertAllPropertiesRead(); [](start = 10, length = 32)

Move all these asserts up into ReadShapeContent so we're not relying on each method to call it.

Comment (simeoncran) [reply source/LottieReader/Serialization/ShapeLayerContents.cs]: Maybe not. The asserts should fire at the point where the method name gives the most context.

---
In reply to: [380911408](https://github.com/windows-toolkit/Lottie-Windows/pull/244#discussion_r380911408) [](ancestors = 380911408)

File: source/LottieReader/Serialization/ShapeLayerContents.cs [hunk 4]
```diff
@@ -446,51 +441,67 @@ TrimPath ReadTrimPath(JCObject obj, in ShapeLayerContent.ShapeLayerContentArgs s
...
+        Repeater ReadRepeater(
+            in LottieJsonObjectElement obj,
+            in ShapeLayerContent.ShapeLayerContentArgs shapeLayerContentArgs)
         {
-            var count = ReadAnimatableFloat(obj.GetNamedObject("c"));
-            var offset = ReadAnimatableFloat(obj.GetNamedObject("o"));
-            var transform = ReadRepeaterTransform(obj.GetNamedObject("tr"), in shapeLayerContentArgs);
+            var count = ReadAnimatableFloat(obj.ObjectOrNullProperty("c"));
+            var offset = ReadAnimatableFloat(obj.ObjectOrNullProperty("o"));
+            var transform = ReadRepeaterTransform(obj.ObjectOrNullProperty("tr").Value, in shapeLayerContentArgs);
```
Comment (simeoncran) [inline source/LottieReader/Serialization/ShapeLayerContents.cs]: >Value [](start = 81, length = 5)

Unchecked.

File: source/LottieReader/Serialization/ShapeLayerContents.cs [hunk 5]
```diff
@@ -446,51 +441,67 @@ TrimPath ReadTrimPath(JCObject obj, in ShapeLayerContent.ShapeLayerContentArgs s
...
-        RepeaterTransform ReadRepeaterTransform(JCObject obj, in ShapeLayerContent.ShapeLayerContentArgs shapeLayerContentArgs)
+        RepeaterTransform ReadRepeaterTransform(
+            in LottieJsonObjectElement obj,
+            in ShapeLayerContent.ShapeLayerContentArgs shapeLayerContentArgs)
         {
-            var startOpacity = ReadOpacityFromObject(obj.GetNamedObject("so", null));
-            var endOpacity = ReadOpacityFromObject(obj.GetNamedObject("eo", null));
+            var startOpacity = ReadOpacityFromObject(obj.ObjectOrNullProperty("so"));
+            var endOpacity = ReadOpacityFromObject(obj.ObjectOrNullProperty("eo"));
             var transform = ReadTransform(obj, in shapeLayerContentArgs);
```
Comment (simeoncran) [inline source/LottieReader/Serialization/ShapeLayerContents.cs]: Add AssertAllPropertiesRead

**Comment by simeoncran** [comment]
>             return new Transform(in shapeLayerContentArgs, anchor, position, scalePercent, rotation, opacity);

Add AssertAllPropertiesRead


---
Refers to: source/LottieReader/Serialization/ShapeLayerContents.cs:559 in 63ceeeb. [](commit_id = 63ceeeb8d611f531c286c156836ef4bcbae9987b, deletion_comment = False)

File: source/LottieReader/Serialization/Animatables.cs [hunk 6]
```diff
@@ -439,19 +509,19 @@ internal void ParseJson(LottieCompositionReader reader, JCObject obj, out IEnume
...
                     }
                 }
             }
 
-            static bool HasKeyframes(JCArray array)
+            static bool HasKeyframes(in LottieJsonArrayElement? array)
             {
-                var firstItem = array[0];
-                return firstItem.Type == JTokenType.Object && firstItem.AsObject().ContainsKey("t");
+                return array?[0].AsObject()?.ContainsProperty("t") == true;
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/Animatables.cs]: >== true [](start = 67, length = 7)

Is this necessary?

Comment (simeoncran) [reply source/LottieReader/Serialization/Animatables.cs]: It's a way of converting the bool? to a bool. If we didn't get a value from array?[0].AsObject()?.ContainsProperty("t") then "null == true" will be false. So it's a way of saying "try and get a bool, but if you can't, default to false".

---
In reply to: [381026970](https://github.com/windows-toolkit/Lottie-Windows/pull/244#discussion_r381026970) [](ancestors = 381026970)

File: source/LottieReader/Serialization/ShapeLayerContents.cs [hunk 7]
```diff
@@ -304,58 +297,63 @@ LinearGradientFill ReadLinearGradientFill(JCObject obj, in ShapeLayerContent.Sha
...
-            var type = SyToPolystarType(obj.GetNamedNumber("sy", double.NaN));
+            var type = SyToPolystarType(obj.DoubleOrNullProperty("sy"));
 
             if (!type.HasValue)
             {
                 return null;
             }
 
-            var points = ReadAnimatableFloat(obj.GetNamedObject("pt"));
+            var polystarType = type.Value;
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/ShapeLayerContents.cs]: >var polystarType = type.Value; [](start = 12, length = 30)

Nit: initialize this just before the `switch` below (line 365); it doesn't appear to be used prior.

File: source/LottieReader/Serialization/ShapeLayerContents.cs [hunk 8]
```diff
@@ -503,38 +547,54 @@ RepeaterTransform ReadRepeaterTransform(JCObject obj, in ShapeLayerContent.Shape
...
+                return ReadTransform(obj.Value, shapeLayerContentArgs);
+            }
+        }
 
-            var positionJson = obj.GetNamedObject("p", null);
+        Transform ReadTransform(
+            in LottieJsonObjectElement obj,
+            in ShapeLayerContent.ShapeLayerContentArgs shapeLayerContentArgs)
+        {
+            // Not clear what these.
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/ShapeLayerContents.cs]: >// Not clear what these. [](start = 12, length = 24)

`Not clear what these are.` || `Not clear whether we need to read these properties.`

File: source/LottieReader/Serialization/Layers.cs [hunk 9]
```diff
@@ -12,36 +12,34 @@ namespace Microsoft.Toolkit.Uwp.UI.Lottie.LottieData.Serialization
...
-            // Not clear whether we need to read these fields.
-            IgnoreFieldThatIsNotYetSupported(obj, "bounds");
-            IgnoreFieldThatIsNotYetSupported(obj, "sy");
-            IgnoreFieldThatIsNotYetSupported(obj, "td");
+            // Not clear whether we need to read these properties.
+            obj.IgnorePropertyThatIsNotYetSupported("bounds", "sy", "td");
 
-            // Field 'hasMask' is deprecated and thus we are intentionally ignoring it
-            IgnoreFieldIntentionally(obj, "hasMask");
+            // Property 'hasMask' is deprecated and thus we are intentionally ignoring it
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/Layers.cs]: >it [](start = 87, length = 2)

.

File: source/LottieReader/Serialization/Layers.cs [hunk 10]
```diff
@@ -197,33 +192,31 @@ void ReadTextData(JCObject obj)
             // "sc":strokeColor
             // "sw":strokeWidth
             // "of":(bool)strokeOverFill
-            IgnoreFieldThatIsNotYetSupported(obj, "d");
-
-            IgnoreFieldThatIsNotYetSupported(obj, "p");
-            IgnoreFieldThatIsNotYetSupported(obj, "m");
+            obj.IgnorePropertyThatIsNotYetSupported("d", "p", "m");
 
             // Array of animatable text properties (fc:fill color, sc:stroke color, sw:stroke width, t:tracking (float))
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/Layers.cs]: >) [](start = 119, length = 1)

Period?

File: source/LottieReader/Serialization/Layers.cs [hunk 11]
```diff
@@ -197,33 +192,31 @@ void ReadTextData(JCObject obj)
...
-        IEnumerable<Mask> ReadMaskProperties(JCArray array)
+        IEnumerable<Mask> ReadMaskProperties(LottieJsonArrayElement array)
         {
             foreach (var elem in array)
             {
-                var obj = elem.AsObject();
+                var obj = elem.AsObject().Value;
 
-                // Ignoring field 'x' because it is not in the official spec
+                // Ignoring property 'x' because it is not in the official spec
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/Layers.cs]: >spec [](start = 75, length = 4)

Period.

File: source/LottieReader/Serialization/LottieJsonElement.cs [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+                }
+            }
+
+            internal LottieJsonArrayElement? AsArray()
+                => _wrapped.Type == JTokenType.Array
+                    ? new LottieJsonArrayElement(_owner, _wrapped)
+                    : (LottieJsonArrayElement?)null;
+
+            internal LottieJsonObjectElement? AsObject()
+            => _wrapped.Type == JTokenType.Object
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/LottieJsonElement.cs]: >=> _wrapped.Type == JTokenType.Object [](start = 12, length = 37)

Indentation.

File: source/LottieReader/Serialization/ShapeLayerContents.cs (hunk 7 above)
Comment (simeoncran) [reply source/LottieReader/Serialization/ShapeLayerContents.cs]: not a nit. That made me realize we can be even more resilient.

---
In reply to: [381038113](https://github.com/windows-toolkit/Lottie-Windows/pull/244#discussion_r381038113) [](ancestors = 381038113)

File: source/LottieReader/Serialization/LottieJsonObjectElement.cs [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+                while (min <= max)
+                {
+                    int mid = (min + max) / 2;
+                    var current = _propertyNames[mid].Item2;
+                    if (propertyName == current)
+                    {
+                        _propertyNames[mid].Item1 = false;
+                        return;
+                    }
+                    else if (string.CompareOrdinal(propertyName, _propertyNames[mid].Item2) < 0)
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/LottieJsonObjectElement.cs]: >_propertyNames[mid].Item2 [](start = 65, length = 25)

`current`

File: source/LottieReader/Serialization/LottieJsonObjectElement.cs [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+                        max = mid - 1;
+                    }
+                    else
+                    {
+                        // Look right.
+                        min = mid + 1;
+                    }
+                }
+
+                // Not found.
```
Comment (derobinsMSFT) [inline source/LottieReader/Serialization/LottieJsonObjectElement.cs]: >// Not found. [](start = 16, length = 13)

Should there be some sort of issue/error reported if we're trying to find a property a given element doesn't have?

Comment (simeoncran) [reply source/LottieReader/Serialization/LottieJsonObjectElement.cs]: No. There are lots of optional values which the JSON doesn't need to specify. Whether a property is needed or not is context dependent.

---
In reply to: [381534955](https://github.com/windows-toolkit/Lottie-Windows/pull/244#discussion_r381534955) [](ancestors = 381534955)

**Pull request merged**